### PR TITLE
[rustdoc] Improve "in parameters" search and search more generally

### LINKED ISSUE #58230: [rustdoc search] Search into parameters might not be as efficient as expected?
To be confirmed.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #58230.

r? @QuietMisdreavus

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
ping from triage @QuietMisdreavus waiting for your review on this

File: src/librustdoc/clean/mod.rs
Comment (QuietMisdreavus) [inline src/librustdoc/clean/mod.rs]: Can you write doc comments for the functions you added? For example, this could be something like "Extract the types from a function's parameters and generics" with maybe a note about trait bounds since you've already added the comment for it.

File: src/librustdoc/clean/mod.rs
Comment (QuietMisdreavus) [inline src/librustdoc/clean/mod.rs]: If you use a `BTreeSet` instead, it comes pre-sorted - otherwise it comes out in an arbitrary order. I'm not sure if you wanted it to be properly sorted or just deduplicated, though.

File: src/librustdoc/clean/mod.rs
Comment (QuietMisdreavus) [inline src/librustdoc/clean/mod.rs]: You can replace this loop with `res.extend(adds)`, since the sets implement the Extend trait.

**Review by QuietMisdreavus — COMMENTED** [review]
Looks good, just a couple small comments. This may clash with https://github.com/rust-lang/rust/pull/59170 whenever that lands.

File: src/librustdoc/clean/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: Ah sure.

File: src/librustdoc/clean/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: Dedup, the order doesn't matter. ;)

**Comment by GuillaumeGomez** [comment]
@QuietMisdreavus I applied your comments.

**Review by QuietMisdreavus — APPROVED** [review]
Excellent, thanks for doing this! This will make the in-doc search much more powerful.

**Comment by eddyb** [comment]
@GuillaumeGomez @QuietMisdreavus Please do not merge PRs that change the compiler (in this case, typeck) without a review from @rust-lang/compiler.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: You didn't have to do any of the typeck changes, the default is already there:
```rust
GenericParamDefKind::Type { ref default, .. } => default.clone(),
```

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: `Option<T>` is not a bound... how does this make sense?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This should be called `Param` or `GenericParam`.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This should be called `is_param` or `is_generic_param`.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This should probably be called `ty`?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This is unchecked recursion, can easily become infinite, probably the cause of #59502.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Same here, unguarded recursion.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Same here, unguarded recursion.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Same here, `Option<A>` is not a bound.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: `bound` here seems to be a misnomer, this is a `param`.

Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: IMO this should be above the `where` handling above, doing it this way is far more confusing.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Why is this by name instead of by `DefId`?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: `ty` is a trait, not a type, so `is_full_generic()` should never be the case.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This name should be `find_concrete_types`, or, even better, `find_type_paths`.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: So this means that `get_real_types(Option<T>)` will look at `T`'s bounds... why?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: `ty` is a trait, so your returned set contains *both* types and traits?

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: The existence of a trait path as a value of `Type` was a mistake, IMO.

If you move `DefId` into `Path`, you can replace
https://github.com/rust-lang/rust/blob/237bf3244fffef501cf37d4bda00e1fce3fcfb46/src/librustdoc/clean/mod.rs#L1196-L1201
With `PolyTrait { trait_: path, generic_params: late_bounds }`.

Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: Oh, you'd need this to take a `Def`, too (which can replace `trait_did`):
https://github.com/rust-lang/rust/blob/237bf3244fffef501cf37d4bda00e1fce3fcfb46/src/librustdoc/clean/mod.rs#L1153-L1154

File: src/librustdoc/clean/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: The point is to replace T by the type/trait if any. For example:

```
fn<T>(x: T)
where T: Display
```

In there, `T` should be replaced with `Display`. In here, it's all about the rustdoc search, not about finding exact types. If we have two types with the same name, we don't care, we just remove one name from the list and that's it.

File: src/librustdoc/clean/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: It was indeed.

File: src/librustdoc/clean/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs]: I don't care about `DefId` in here, I only care about names. Don't forget it's to generate rustdoc search array. I'm not looking for a given type but for a given type **name**.

Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: You're searching for a type parameter though, not a struct/enum.
And you implemented the other search for type parameters by `DefId`, so it's inconsistent.

File: src/librustdoc/clean/mod.rs
Comment (eddyb) [reply src/librustdoc/clean/mod.rs]: Sure, but you can get the desired effect without using confusing terminology.

So what you want is "types of arguments" *and* "bounds on parameters in those arguments"?

Why bother with that complication tho? Why not just record all paths in a signature, be it args/return, or bounds/where clauses? Getting all that is less work and much easier to reason about.

**Comment by eddyb** [comment]
Opened #59789.

**Pull request merged**
